File: 2025-1/sesion-de-clase-18/SoftProgNet/SoftProgPersistencia/DAOImpl/Clientes/ClienteDAOImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PUCP.Edu.Pe.SoftProg.Config;
using PUCP.Edu.Pe.SoftProg.Modelo.Clientes;
using PUCP.Edu.Pe.SoftProg.Modelo.Logistica.Almacen;
using PUCP.Edu.Pe.SoftProg.Modelo.RRHH;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO.Clientes;

namespace PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl.Clientes {
    public class ClienteDAOImpl : BaseDAOImpl<Cliente>, IClienteDAO {
        protected override DbCommand CommandoInsertar(DbConnection conn, Cliente cliente) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarCliente";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.Double, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cliente.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand CommandoInsertar(DbConnection conn, DbTransaction transaccion, Cliente cliente) {
            DbCommand cmd = this.CommandoInsertar(conn, cliente);
            cmd.Transaction = transaccion;

            return cmd;
        }

        protected override DbCommand CommandoModificar(DbConnection conn, Cliente cliente) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarCliente";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.Double, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cliente.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, cliente.Id);

            return cmd;
        }

        protected override DbCommand CommandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarCliente";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoBuscar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarClientePorId";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoListar(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarClientes";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            return cmd;
        }

        protected override Cliente mapearModelo(DbDataReader reader) {
            return new Cliente {
                Id = Convert.ToInt32(reader["id"]),
                Dni = Convert.ToString(reader["dni"]),
                Nombre = Convert.ToString(reader["nombre"]),
                ApellidoPaterno = Convert.ToString(reader["apellidoPaterno"]),
                Genero = Convert.ToChar(reader["genero"]),
                FechaNacimiento = Convert.ToDateTime(reader["fechaNacimiento"]),
                Categoria = (Categoria)Enum.Parse(typeof(Categoria), Convert.ToString(reader["categoria"])),
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }

        public Cliente BuscarPorDni(string dni)
        {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.ComandoBuscarPorDni(conn, dni))
                using (DbDataReader reader = cmd.ExecuteReader()) {
                    if (!reader.HasRows) {
                        Console.Error.WriteLine("No se encontro el registro con DNI: " + dni);
                        return default;
                    }

                    reader.Read();
                    return this.mapearModelo(reader);
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la busueda: " + e.Message);
                throw new InvalidOperationException("No se pudo buscar el registro.", e);
            }
        }

        private DbCommand ComandoBuscarPorDni(DbConnection conn, string dni)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarClientePorDni";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.Int32, dni);

            return cmd;
        }
    }
}

File: 2025-1/sesion-de-clase-23/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/clientes/ClienteDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.clientes;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.config.DBManager;
import pe.edu.pucp.inf30.softprog.dao.clientes.IClienteDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAOImpl;
import pe.edu.pucp.inf30.softprog.model.clientes.Cliente;
import pe.edu.pucp.inf30.softprog.model.clientes.Categoria;

/**
 *
 * @author eric
 */
public class ClienteDAOImpl extends BaseDAOImpl<Cliente> implements IClienteDAO {
    @Override
    protected CallableStatement comandoInsertar(Connection conn, Cliente modelo) throws SQLException {
//        String call = getPrefijoLlamada();
        String sql = "{CALL insertarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", new java.sql.Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_categoria", modelo.getCategoria().name());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }

    @Override
    protected CallableStatement comandoModificar(Connection conn, Cliente modelo) throws SQLException {
        String sql = "{CALL modificarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", new java.sql.Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_categoria", modelo.getCategoria().name());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        return cmd;
    }

    @Override
    protected CallableStatement comandoEliminar(Connection conn, int id) throws SQLException {
        String sql = "{CALL eliminarCliente(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoBuscar(Connection conn, int id) throws SQLException {
        String sql = "{CALL buscarClientePorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoListar(Connection conn) throws SQLException {
        String sql = "{CALL listarClientes()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }

    @Override
    protected Cliente mapearModelo(ResultSet rs) throws SQLException {
        Cliente cliente = new Cliente();
        cliente.setId(rs.getInt("id"));
        cliente.setDni(rs.getString("dni"));
        cliente.setNombre(rs.getString("nombre"));
        cliente.setApellidoPaterno(rs.getString("apellidoPaterno"));
        cliente.setGenero(rs.getString("genero").charAt(0));
        cliente.setFechaNacimiento(rs.getTimestamp("fechaNacimiento"));
        cliente.setCategoria(Categoria.valueOf(rs.getString("categoria")));
        cliente.setActivo(rs.getBoolean("activo"));
        return cliente;
    }

    protected CallableStatement comandoBuscarPorDni(Connection conn, String dni) throws SQLException {
        String sql = "{CALL buscarClientePorDni(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_dni", dni);
        return cmd;
    }
    
    @Override
    public Cliente buscarPorDni(String dni) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            CallableStatement cmd = this.comandoBuscarPorDni(conn, dni);
        ) {
            ResultSet rs = cmd.executeQuery();
            
            if (!rs.next()) {
                System.err.println("No se encontro el registro con dni: " + dni);
                return null;
            }
            
            return this.mapearModelo(rs);
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la insercion: " + e.getMessage());
            throw new RuntimeException("No se pudo insertar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al insertar el registro.", e);
        }
    }
}
